libsync/TreeTopology.cpp

```diff
@@ -0,0 +1,212 @@
+}
+
+/**
+ * @brief : select child nodes of given node from peers recursively
+ *          if the any child node doesn't exist in the peers, select the grand child nodes, etc.
+ * @params :
+ *  1. _selectedNodeList: return value, the selected child nodes(maybe the grand child, etc.)
+ *  2. _parentIndex: the index of the node who needs select child nodes from the given peers
+ *  3. _peers: the nodeIDs of peers maintained by syncStatus
+ */
+void TreeTopology::recursiveSelectChildNodes(std::shared_ptr<h512s> _selectedNodeList,
+    ssize_t const& _parentIndex, std::shared_ptr<std::set<dev::h512>> _peers,
+    int64_t const& _startIndex)
+{
+    // if the node doesn't locate in the group
+    dev::h512 selectedNode;
+    for (ssize_t i = 0; i < m_treeWidth; i++)
+    {
+        ssize_t expectedIndex = _parentIndex * m_treeWidth + i + 1;
+        // when expectedIndex is bigger than m_currentConsensusNodes.size(), return
+        if (expectedIndex > m_endIndex)
+        {
+            break;
+        }
+        auto selectedIndex = getSelectedNodeIndex(expectedIndex, _startIndex);
+        if (!getNodeIDByIndex(selectedNode, selectedIndex))
+        {
+            continue;
+        }
+        // the child node exists in the peers
+        if (_peers->count(selectedNode))
+        {
+            _selectedNodeList->push_back(selectedNode);
+        }
+        // the child node doesn't exist in the peers, select the grand child recursively
+        else
+        {
+            if (expectedIndex < m_endIndex)
+            {
+                recursiveSelectChildNodes(_selectedNodeList, expectedIndex, _peers, _startIndex);
+            }
+        }
+        // the last node
+        if (expectedIndex == m_endIndex)
+        {
+            break;
+        }
+    }
+}
+
+/**
+ * @brief : select parent node of given node from peers
+ *          if the parent node doesn't exist in the peers, select the grand parent, etc.
+ *
+ * @params :
+ *  1. _selectedNodeList: return value, the selected parent node(maybe the grand parent, etc.)
+ *  2. _peers: the nodeIDs of peers maintained by syncStatus
+ *  3. _nodeIndex: the index of the node that need select parent from given peers
+ */
+void TreeTopology::selectParentNodes(std::shared_ptr<dev::h512s> _selectedNodeList,
+    std::shared_ptr<std::set<dev::h512>> _peers, int64_t const& _nodeIndex,
+    int64_t const& _startIndex)
+{
+    ssize_t parentIndex = (_nodeIndex - 1) / m_treeWidth;
+    // the parentNode is the node-slef
+    if (parentIndex == _nodeIndex)
+    {
+        return;
+    }
+    dev::h512 selectedNode;
+    while (parentIndex > 0)
+    {
+        // find the parentNode from the peers
+        auto selectedIndex = getSelectedNodeIndex(parentIndex, _startIndex);
+        if (getNodeIDByIndex(selectedNode, selectedIndex) && _peers->count(selectedNode))
+        {
+            _selectedNodeList->push_back(selectedNode);
+            break;
+        }
+        if (parentIndex <= 0)

```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Opposite inner 'if' condition leads to a dead code block.](https://app.codacy.com/app/fisco/FISCO-BCOS/pullRequest?prid=4731778)